### PR TITLE
pkgs.dockertools.buildLayeredImage: support custom layer packing strategies

### LINKED ISSUE #48462: dockerTools.buildLayeredImage creates skewed sized layers
## Issue description
Recently `dockerTools.buildLayeredImage` ( https://grahamc.com/blog/nix-and-layered-docker-images ) functionality was merged into master. 

It creates docker images with layers of a size with a distribution that is very skewed. See below for more information.

### Steps to reproduce
```
let
  old_pkgs = import <nixpkgs> {};
  pkgs  = <snip (assumes that it points at master> ;
 inherit (pkgs) stdenv;

in pkgs.dockerTools.buildLayeredImage {
   contents =  [pkgs.bashInteractive old_pkgs.curl old_pkgs.jq old_pkgs.emacs old_pkgs.git
                old_pkgs.cacert old_pkgs.openssh old_pkgs.gnugrep
                old_pkgs.coreutils
                old_pkgs.which
                old_pkgs.nix
                old_pkgs.gnused old_pkgs.file old_pkgs.gawk];

  name = "foo/bar";
  tag = "latest";
  config.Cmd = [ "/bin/bash" ];
  maxLayers = 25;
}
```

The layers looks like: 
```
b6cdec945d1b: Loading layer [==================================================>]    477MB/477MB
86e60f3ed08f: Loading layer [==================================================>]  174.1kB/174.1kB
79d24adb2c31: Loading layer [==================================================>]  348.2kB/348.2kB
9354bd19facd: Loading layer [==================================================>]  12.19MB/12.19MB
611dae553d72: Loading layer [==================================================>]  1.178MB/1.178MB
cb916eb69cd1: Loading layer [==================================================>]  71.68kB/71.68kB
ee38b4411beb: Loading layer [==================================================>]    1.7MB/1.7MB
05bad8e60f92: Loading layer [==================================================>]    809kB/809kB
d30df3eb574c: Loading layer [==================================================>]    768kB/768kB
bb3ba7943090: Loading layer [==================================================>]  40.96kB/40.96kB
e7e6d2f94b2e: Loading layer [==================================================>]  5.612MB/5.612MB
15048416f85e: Loading layer [==================================================>]  1.526MB/1.526MB
b00ca1c01fd7: Loading layer [==================================================>]    512kB/512kB
0518754facbc: Loading layer [==================================================>]  92.16kB/92.16kB
13859dfd2c8b: Loading layer [==================================================>]  266.2kB/266.2kB
e696a9248983: Loading layer [==================================================>]    256kB/256kB
560c99588b76: Loading layer [==================================================>]  81.92kB/81.92kB
dcf8aeee86d0: Loading layer [==================================================>]   51.2kB/51.2kB
9b58d0959ca7: Loading layer [==================================================>]  2.898MB/2.898MB
9caed6f91ab9: Loading layer [==================================================>]  112.6kB/112.6kB
e7379e4a6dcb: Loading layer [==================================================>]  1.352MB/1.352MB
4a5ae92f18dc: Loading layer [==================================================>]  40.96kB/40.96kB
d91816088e26: Loading layer [==================================================>]  40.96kB/40.96kB
7ddbb507422c: Loading layer [==================================================>]  21.79MB/21.79MB
```

As you can see the size is unreasonably distributed (a 477MB layer with a bunch of tiny ones). 

My question is: is this the intended behavior? If not, can we have tests that make sure the distribution of the size of the layers follows a particular (statistical) distribution.

## Relevant people
@graham-at-target 

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.74, NixOS, 18.09.git.08005ed (Jellyfish)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.1.1`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

### PR DESCRIPTION
**Pull request by bhpdt**
###### Motivation for this change

The default algorithm does not work particularly well for some use cases, such
as applications with very large dependency closures when we want to optimize
iterative releases of the same application container rather than optimize across
different application containers.

Currently there's limited sharing, since almost all of the packages are in the
final layer along with the app itself. This commit adds a new layering strategy
which will emphasize cache hits for the application itself and the dependencies
closest to the application, which are the most likely to change, while leaving
the more-likely-to-be-constant low level dependencies smooshed into the first layer.

Since the current algorithm likely works well for many use cases, this leaves
the default unchanged, but enhances `buildLayeredImage` to be able to take a
per-nix-expression custom layering algorithm as a parameter.

Resolves #48462

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [X] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bhpdt** [comment]
@ofborg test docker-tools

**Comment by grahamc** [comment]
Looks pretty cool.

**Comment by roberth** [comment]
The popularity sorting should be part of the pluggable strategy. @adrian-gierakowski is working on a strategy that's based on iterative closure splitting instead of the popularity algorithm. See https://github.com/NixOS/nixpkgs/issues/48462#issuecomment-794179817

**Comment by adrian-gierakowski** [comment]
@bhpdt as @roberth suggested, I believe we might want to slice the code in different place than what you've done in this PR, as people might want to use strategies which do not rely on popularity sorting. Currently the code structure looks like this (pseudo nix like code):

```
pathsSortedByPopularity = referencesByPopularity overallClosure
pathsWithUnnecesaryDrvsRemoved = removeUnnecesaryPath unnecessaryDrvs pathsSortedByPopularity
layerSpec = layerStrategy availableLayerCount pathsWithUnnecesaryDrvsRemoved
```

where `pathsSortedByPopularity` is an array of paths and `layerSpec` is an array of array of paths

Your implementation assumes that users always want to sort by popularity and only allows to customise how the `pathsSortedByPopularity` are packaged into layers. However It would be more flexible if a custom strategy would operate on `overallClosure` and output the final `layerSpec`. So I think the code in `streamLayeredImage` should look more like:

```
layerSpec = layerStrategy availableLayerCount unnecessaryDrvs overallClosure
```

or even:

```
layerSpec = layerStrategy availableLayerCount baseJson contentsList
```

File: pkgs/build-support/docker/layer-strategies/popularity_weighted_top.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import json
+import sys
+
+available_layers = int(sys.argv[1])
+pkgs = [x.strip() for x in sys.stdin.readlines()]
+
+# Layers [2, limit] have 1 pkg each, of the user's top-most pkgs,
+# and layer 1 has the remainder.
+user_pkgs = pkgs[-(available_layers - 1):]
+layers = [pkgs[:-(available_layers - 1)]]
```
Comment (lbpdt) [inline pkgs/build-support/docker/layer-strategies/popularity_weighted_top.py:10]: Just realized this produces an empty layer when there are less `pkgs` than `available_layers`: `layers = [ [] ]`.

You could check if you have any base layers before initializing `layers`. Here's what I tested locally:
```python
base_pkgs = pkgs[:-(available_layers - 1)]

layers = [base_pkgs] if base_pkgs else []
layers.extend([pkg] for pkg in user_pkgs)
```

Comment (roberth) [reply pkgs/build-support/docker/layer-strategies/popularity_weighted_top.py:10]: We should have a regression test for this.

**Pull request closed without merging**